sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/variant/variantExpressions.scala

```diff
@@ -63,3 +70,300 @@ case class ParseJson(child: Expression) extends UnaryExpression
+    }
+
+  // Parse key segment like `.name`, `['name']`, or `["name"]`.
+  private def key: Parser[PathSegment] =
+    for {
+      key <- '.' ~> "[^\\.\\[]+".r | "['" ~> "[^\\'\\?]+".r <~ "']" |
+        "[\"" ~> "[^\\\"\\?]+".r <~ "\"]"
+    } yield {
+      PathSegment(key, 0)
+    }
+
+  private val parser: Parser[List[PathSegment]] = phrase(root ~> rep(key | index))
+
+  def parse(str: String): Option[Array[PathSegment]] = {
+    this.parseAll(parser, str) match {
+      case Success(result, _) => Some(result.toArray)
+      case _ => None
+    }
+  }
+}
+
+/**
+ * The implementation for `variant_get` and `try_variant_get` expressions. Extracts a sub-variant
+ * value according to a path and cast it into a concrete data type.
+ * @param child The source variant value to extract from.
+ * @param path A literal path expression. It has the same format as the JSON path.
+ * @param schema The target data type to cast into.
+ * @param failOnError Controls whether the expression should throw an exception or return null if
+ *                    the cast fails.
+ * @param timeZoneId A string identifier of a time zone. It is required by timestamp-related casts.
+ */
+case class VariantGet(
+    child: Expression,
+    path: Expression,
+    schema: DataType,
+    failOnError: Boolean,
+    timeZoneId: Option[String] = None)
+    extends BinaryExpression
+    with TimeZoneAwareExpression
+    with NullIntolerant
+    with ExpectsInputTypes
+    with CodegenFallback
+    with QueryErrorsBase {
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val check = super.checkInputDataTypes()
+    if (check.isFailure) {
+      check
+    } else if (!path.foldable) {
+      DataTypeMismatch(
+        errorSubClass = "NON_FOLDABLE_INPUT",
+        messageParameters = Map(
+          "inputName" -> toSQLId("path"),
+          "inputType" -> toSQLType(path.dataType),
+          "inputExpr" -> toSQLExpr(path)
+        )
+      )
+    } else if (!VariantGet.checkDataType(schema)) {
+      DataTypeMismatch(
+        errorSubClass = "CAST_WITHOUT_SUGGESTION",
+        messageParameters = Map(
+          "srcType" -> toSQLType(VariantType),
+          "targetType" -> toSQLType(schema)
+        )
+      )
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override lazy val dataType: DataType = schema.asNullable
+
+  @transient private lazy val parsedPath = {
+    val pathValue = path.eval().toString
+    VariantPathParser.parse(pathValue).getOrElse {
+      throw QueryExecutionErrors.invalidVariantGetPath(pathValue, prettyName)
+    }
+  }
+
+  final override def nodePatternsInternal(): Seq[TreePattern] = Seq(VARIANT_GET)
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(VariantType, StringType)
+
+  override def prettyName: String = if (failOnError) "variant_get" else "try_variant_get"
+
+  override def nullable: Boolean = true
+
+  protected override def nullSafeEval(input: Any, path: Any): Any = {
```

Comment:
Some ideas to simplify the implementation:

Add java util methods for the implementation, e.g.
```
public Object getValueFromVariant(VariantVal variant, String path) {
  ...
}
```

We can add more overloads to avoid boxing
```
public boolean getBooleanFromVariant(VariantVal variant, String path) {
  ...
}
```

Then, `VariantGet` can extend `RuntimeReplaceable` and use `StaticInvoke` to call the java util methods, and also `Cast` to do the cast work
```
case class VariantGet ...
  lazy val replacement: Expression = Cast(StaticInvoke(...), targetType, ansiEnabled = failOnError)
```

Reply:
There is an issue with this approach: we cannot know the return type of `StaticInvoke(...)` in advance. It is possible that there are two variant rows with different content types and they can both be cast into the same type. E.g., both `variant_get(parse_json('1'), '$', 'int')` and `variant_get(parse_json('"1"'), '$', 'int')` should return `1`. If the `StaticInvoke(...)` must have a return type, it can only be the variant type. We still need the current `VariantGet.cast` function as the internal implementation of `Cast(Variant, targetType)`, so it doesn't really simplify anything.

Reply:
BTW, we can probably wait for https://github.com/apache/spark/pull/45714 to add the codegen utils for variant type.

Reply:
There is yet another reason against using `StaticInvoke`. The path parameter must be an literal, and I can make use of this requirement to avoid repeated path parsing. However, I cannot find how to do a similar caching in `StaticInvoke`.

Using `StaticInvoke` won't simplify the current implementation. It can indeed simplify the implementation if we want to support native codegen rather than depending on `CodegenFallback`. I think that is an optional optimization we can do in the future, when we can manually write the codegen for `VariantGet`.

Reply:
You can pass anything to `StaticInvoke`, including arbitrary java object, using `Literal` with `ObjectType`. I'm against writing codegen by hand, as it's hard to debug, and error-prone (maybe inconsistent with the interpreted implementation).

Reply:
I didn't mean writing everything by hand. Essentially, we create a method that implements `VariantGet`, and the class only needs some boilerplate code to call this method (similar to the code in `StaticInvoke` itself).

There is still another reason why I don't like `StaticInvoke`. In the future, I will write some optimizer rules on `VariantGet` (e.g., to push it down a join). This is why I added a new `TreePattern` ``VARIANT_GET` in this PR. The optimizer rule will run after `RuntimeReplaceable` expression is replaced, so it will become `StaticInvoke` and no longer has this tree pattern, and the optimizer rule can no longer prune expressions. Plus, matching against `StaticInvoke` is also more complex than matching against `VariantGet`.

Reply:
I have done some experiments with the `StaticInvoke` approach. Suppose I have encapsulated the `VariantGet` implementation into the following function:
```
case object VariantGetCodegen {
  def variantGet(input: VariantVal, parsedPath: Array[PathSegment],
                 dataType: DataType, failOnError: Boolean, zoneId: Option[String]): Any = {...}
}
```
and make `VariantGet` a `RuntimeReplaceable` expression with a replacement of `StaticInvoke` that invokes `VariantGetCodegen.variantGet`. It still won't directly work because the codegen logic of `StaticInvoke` assumes the return type of the method directly matches the return type, but the return type of `VariantGetCodegen.variantGet` is `Any`.

In order to make it work, I have to create a wrapper for each return type, like:

```
case object VariantGetCodegen {
  def variantGetByte(input: VariantVal, parsedPath: Array[PathSegment],
                 dataType: DataType, failOnError: Boolean, zoneId: Option[String]): Byte =
    variantGet(input, parsedPath. dataType, failOnError, zoneId).asInstanceOf[Byte]
  def variantGetShort(input: VariantVal, parsedPath: Array[PathSegment],
                 dataType: DataType, failOnError: Boolean, zoneId: Option[String]): Short =
    variantGet(input, parsedPath. dataType, failOnError, zoneId).asInstanceOf[Short]
  def variantGetStruct(input: VariantVal, parsedPath: Array[PathSegment],
                 dataType: DataType, failOnError: Boolean, zoneId: Option[String]): InternalRow =
    variantGet(input, parsedPath. dataType, failOnError, zoneId).asInstanceOf[InternalRow]
  ...
}
```

and pick one method according to the return type. It is very cumbersome and doesn't really avoid any boxing/unboxing costs.

On the other hand, if we have this `VariantGetCodegen.variantGet` method, it is reasonably easy to write the codegen by hand. I just need to cast the return value of this method into the desired type. The whole point of using `StaticInvoke` is to simplify the implementation, but I think it actually makes the implementation much more complex.

Reply:
In case you are interested, I have a draft for the manual codegn version. I think I can add it in a follow-up PR. Personally, I don't feel the code has any red flag, and it is much better than the `StaticInvoke` approach.

```
  protected override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
    val childCode = child.genCode(ctx)
    val tmp = ctx.freshVariable("tmp", classOf[Object])
    val parsedPathArg = ctx.addReferenceObj("parsedPath", parsedPath)
    val dataTypeArg = ctx.addReferenceObj("dataType", dataType)
    val zoneIdArg = ctx.addReferenceObj("zoneId", timeZoneId)
    val code = code"""
      ${childCode.code}
      boolean ${ev.isNull} = ${childCode.isNull};
      ${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
      if (!${ev.isNull}) {
        Object $tmp = org.apache.spark.sql.catalyst.expressions.variant.VariantGet.variantGet(
          ${childCode.value}, $parsedPathArg, $dataTypeArg, $failOnError, $zoneIdArg);
        if ($tmp == null) {
          ${ev.isNull} = true;
        } else {
          ${ev.value} = (${CodeGenerator.boxedType(dataType)})$tmp;
        }
      }
    """
    ev.copy(code = code)
  }
```

Reply:
oh dynamic return type is indeed an issue. Your proposal LGTM: we can still put a large chunk of implementation code in a static method and then call it in both the interpreted and codegen versions.